src/test/chain_tests.cpp

```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+                                    0b10101000);
+    // Odd values: the 2nd and 3rd bits are zeroed
+    // Test with various odd values with at least 4 bits set
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b10010011),
+                                    0b10000001);
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b10100011),
+                                    0b10000001);
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b11000011),
+                                    0b10000001);
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b10010101),
+                                    0b10000001);
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b10011001),
+                                    0b10000001);
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b10101011),
+                                    0b10100001);
+    // Some longer random values (even and odd)
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b0001011101101000),
+                                    0b0001011101100000);
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b0001011101101001),
+                                    0b0001011101000001);
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b0110101101011000),
+                                    0b0110101101010000);
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b0110101101011001),
+                                    0b0110101101000001);
+    // All values 0-10
+    BOOST_CHECK_EQUAL(GetSkipHeight(0), 0);
+    BOOST_CHECK_EQUAL(GetSkipHeight(1), 0);
+    BOOST_CHECK_EQUAL(GetSkipHeight(2), 0);
+    BOOST_CHECK_EQUAL(GetSkipHeight(3), 1);
+    BOOST_CHECK_EQUAL(GetSkipHeight(4), 0);
+    BOOST_CHECK_EQUAL(GetSkipHeight(5), 1);
+    BOOST_CHECK_EQUAL(GetSkipHeight(6), 4);
+    BOOST_CHECK_EQUAL(GetSkipHeight(7), 1);
+    BOOST_CHECK_EQUAL(GetSkipHeight(8), 0);
+    BOOST_CHECK_EQUAL(GetSkipHeight(9), 1);
+    BOOST_CHECK_EQUAL(GetSkipHeight(10), 8);
+}
+
+BOOST_AUTO_TEST_CASE(skip_height_analysis)
+{
+    // This test case indirectly tests the `GetSkipHeight` function,
+    // and asserts that its properties are suitable for performant
+    // ancestor search.
+    // It does so by imitating the `GetAncestor` function to
+    // look for an ancestor, and ensuring that an arbitrary
+    // ancestor is found in a number of steps that is much lower
+    // than simple linear traversal.
+
+    // Build a chain
+    const auto chain_size{200'000};
+    std::vector<std::unique_ptr<CBlockIndex>> block_index;
+    block_index.reserve(chain_size);
+    // Create genesis block
+    block_index.push_back(std::make_unique<CBlockIndex>());
+    for(auto i{1}; i < chain_size; ++i) {
+        auto new_index{std::make_unique<CBlockIndex>()};
+        new_index->pprev = block_index.back().get();
+        BOOST_CHECK(new_index->pprev);
+        new_index->nHeight = new_index->pprev->nHeight + 1;
+        new_index->BuildSkip();
+        block_index.push_back(std::move(new_index));
+    }
+
+    const auto runs{1000};
+    const auto min_distance{500};
+    FastRandomContext ctx(/*fDeterministic =*/true);
```

Comment:
making it deterministic kinda' defeats the purpose in my opinion (it's a low bar to construct something that accidentally works, a lot more difficult to find the exact properties of a system that work less predictable values)

nit: extra space before `=`